### PR TITLE
Print provider-metadata.json files per domain

### PR DESCRIPTION
**Pull request by immqu**
This PR addresses #520. It adds an enumerate method to the providermetaloader.go which loads the PMDs from the well-known path, the security.txt, or the DNS URL (as defined in the spec in [requirement 7.3.1](https://docs.oasis-open.org/csaf/csaf/v2.0/os/csaf-v2.0-os.html#731-finding-provider-metadatajson)).

The PR also includes usage of the new method in the csaf_downloader and its main method.

There is, however, still a doubt that I have about the issue: Why does the load method compare the PMD from the well-known path to the first one found in the security.txt? I couldn't find a respective requirement in the spec. It could be the case that a provider provides the PMD in the well-known path and only advertises additional PMDs in the security.txt, right?

Looking forward to your feedback!

### CODE REVIEW

File: cmd/csaf_downloader/config.go [hunk 1]
```diff
@@ -58,6 +58,8 @@ type config struct {
 	IgnorePattern        []string          `long:"ignore_pattern" short:"i" description:"Do not download files if their URLs match any of the given PATTERNs" value-name:"PATTERN" toml:"ignore_pattern"`
 	ExtraHeader          http.Header       `long:"header" short:"H" description:"One or more extra HTTP header fields" toml:"header"`
 
+	EnumeratePMDOnly bool `long:"enumerate_pmd_only" description:"If this flag is set to true, the donwloader will only enumerate valid provider metadata files, but not download documents" toml:"enumerate_pmd_only"`
```
Comment (bernhardreiter) [inline cmd/csaf_downloader/config.go:61]: Typo "donwloader"

**Comment by bernhardreiter** [comment]
Hi @immqu, thanks for your contribution!

We will review it in the next days. I've started the automatic checks.
Note that I found a typo and the linter also found something (see check report).

**Comment by immqu** [comment]
Hi, I have fixed the typo and linting errors.

Also, I realized the doubt I described above about the Load method came from overlooking one sentence in the spec ("If one of the URLs fulfills requirement 9, this MUST be used as the first CSAF entry in the security.txt"). Currently, the new Enumerate method lists all the PMDs including the duplicated one from the security.txt, but I think we could leave it like that, since this allows to verify that the PMDs are provided correctly.

**Comment by bernhardreiter** [comment]
@immqu thanks! Can you run the Integration Test action on your branch for an additional check? (Go to Actions and then run it manually on your branch.)

File: csaf/providermetaloader.go [hunk 2]
```diff
@@ -108,8 +109,51 @@ func NewProviderMetadataLoader(client util.Client) *ProviderMetadataLoader {
 	}
 }
 
-// Load loads a provider metadata for a given path.
-// If the domain starts with `https://` it only attemps to load
```
Comment (bernhardreiter) [inline csaf/providermetaloader.go:112]: I believe the comment for documentation "If the domain starts with `https://` it only attemps to load this as PMD" or so, should stay in.

**Comment by immqu** [comment]
I have updated the fork with the new changes and successfully run the integration tests!

**Comment by bernhardreiter** [comment]
docs/csaf_downloader.md needs an update for the new flag (it is done manually after running --help)

**Review by bernhardreiter — APPROVED** [review]
works and LGTM

(I did check the autodoc updates, though I wonder why this has not done with our commits in the main repository, but it should be an improvement if it does in this way.)

There is improvement potential for subsequent changes, like removing duplicates or changing the name of the function to something that removes the "numbering" part and potentially to merge the location functions to use one implementation of the search strategy.

File: csaf/providermetaloader.go [hunk 3]
```diff
@@ -108,8 +109,51 @@ func NewProviderMetadataLoader(client util.Client) *ProviderMetadataLoader {
...
+	secResults := pmdl.loadFromSecurity(domain)
+	slog.Info("Found provider metadata results in security.txt", "num", len(secResults))
+
+	for _, result := range secResults {
+		if result.Valid() {
+			resPMDs = append(resPMDs, result)
+		}
+	}
+
+	// According to the spec, only if no PMDs have been found, the should DNS URL be used
```
Comment (ThomasJunk) [inline csaf/providermetaloader.go:146]: This looks like a typo in comment.

Perhaps it should read

>[...]then the DNS URL should be used

or

>[...] the DNS URL should be used

**Review by ThomasJunk — APPROVED** [review]
LGTM.

Typo in comment should not block merging.

**Pull request merged**